### ISSUE TITLE
bf16 + peak head freezes the loss and zeroes the gradient

### ISSUE THREAD

**Issue by mehdiforoozandeh** [mehdiforoozandeh/CANDII#1]
## Summary

With `--heads count,peak --precision bf16`, the peak-head BCE term evaluates `log(0)` and the loss
becomes `-inf`. Training cannot recover from this.

The bug exists only in the **combination** of the two features. Each is correct on its own. They were
developed on separate branches from the same base and neither could see the other, so no test on
either branch covers the configuration.

## Mechanism

`decoder.PeakHead.forward` returns a probability:

```python
return torch.sigmoid(self.linear_peak(feat)).squeeze(-1)
```

and `train.peak_bce_loss` guards it with `clamp(1e-6, 1 - 1e-6)` before taking logs.

Both steps fail in bf16, for the same reason the NB head needed an fp32 fence:

1. `sigmoid(logit)` in bf16 **saturates to exactly 1.0** by a logit of about 8. In fp32 the same
   logit gives 0.99966. A logit of 8 is an ordinary confident peak call, not a tail case.
2. The clamp cannot rescue it, because `1 - 1e-6` **is not representable in bf16 as anything other
   than 1.0**. The upper bound collapses onto the value it is meant to exclude.

So `log(1 - p) = log(0) = -inf`.

Verified directly:

```
logit=8.0   bf16 sigmoid=1.0  after clamp=1.0  log(1-p)=-inf
logit=12.0  bf16 sigmoid=1.0  after clamp=1.0  log(1-p)=-inf
logit=20.0  bf16 sigmoid=1.0  after clamp=1.0  log(1-p)=-inf
```

This is the same hazard the bf16 work already identified for the NB `p`, where
`probs.clamp(eps, 1-eps)` clamping onto exactly `1.0` puts the value outside `NegativeBinomial`'s
support. Confirmed independently: `bfloat16(1 - 1e-6) == 1.0` is `True`, and
`NegativeBinomial(total_count=5.0, probs=1.0)` raises `ValueError`.

## The related, quieter defect in the signal head

`GaussianSignalHead` computes `var = softplus(...) + 1e-6`. In bf16 the `+ 1e-6` **changes nothing**
at any tested value — bf16's steps are far coarser than `1e-6`:

```
x=0.5   bf16 var=0.97265625  floor applied: False   fp32 var=0.97407800
x=1.0   bf16 var=1.31250000  floor applied: False   fp32 var=1.31326258
x=5.0   bf16 var=5.00000000  floor applied: False   fp32 var=5.00671625
```

The floor exists to keep the NLL's `(y - mu)^2 / var` division safe. Under bf16 it is simply absent.
This does not currently produce `-inf` because softplus outputs sit near 1, but the protection is
gone exactly where it would matter — a saturating-negative logit driving `var` toward 0.

## Why no test caught it

The fp32 fence in `decoder.SymmetricDecoder.forward` upcasts `eta`, `raw_n`, `y_meta` and `log_ref`.
It does **not** upcast `feat`, which is what both optional heads read. The fence was written when the
optional heads did not exist.

## Fix — pick one

**Option A (minimal).** Move the `head_signal` / `head_peak` calls inside the existing
`fp32_fence(...)` block in `SymmetricDecoder.forward`, adding `feat` to the fenced tensors. Same
reasoning already applied to `p`.

**Option B (preferred).** Change `PeakHead` to emit **logits** and switch the loss to
`binary_cross_entropy_with_logits`. This removes the hazard at its source rather than fencing around
it, and it is numerically better at any precision. `PeakHead`'s own docstring already says
"A logit head would be numerically better and would not be this head" — it emits a probability only
to match the production interface. Evaluation consumes a probability either way, so
`sigmoid()` at the read site keeps the external contract. Option B should still fence the signal
head's variance floor.

## Acceptance criteria

- A test builds a model with `heads=("count","peak")`, runs a forward+loss under bf16 autocast with
  at least one logit >= 8, and asserts the loss is finite.
- A test asserts the signal head's variance floor is actually applied under bf16 (`var > softplus`).
- The test fails if the fence (or the logit head) is removed — assert against an unfenced reference,
  not only for finiteness.
- `tools/golden.py check` still reports 0 ULP at the defaults.
- The full suite passes.

**Comment by mehdiforoozandeh** [comment]
## Correction: the failure mode in this issue was wrong, and the real one is worse

The original report said the loss goes `-inf`. **It does not.** `binary_cross_entropy` clamps its
internal log at `-100`, so the forward stays finite and nothing raises. I checked `log(1-p)` in
isolation, saw `-inf`, and did not check what the loss function actually returns.

What the old path really did, measured on a negative label under bf16, with the backward:

| logit | bf16 loss | bf16 d(loss)/d(logit) | fp32 loss | fp32 gradient |
|---|---|---|---|---|
| 4 | 3.9357 | 0.9805 | 4.0182 | 0.9820 |
| 6 | 5.5452 | 0.9961 | 6.0025 | 0.9975 |
| **8** | **13.8023** | **0** | 8.0004 | 0.9997 |
| **12** | **13.8023** | **0** | 11.9911 | 0.99999 |
| **40** | **13.8023** | **0** | 13.8023 | 0 |

The loss **freezes at 13.8023** — identical for logits 8, 12 and 40, so every confidence above the
cliff is indistinguishable — and the **gradient is exactly zero**. The peak head silently stops
learning from precisely the positions it is most confidently wrong about, and the loss curve stays
plausible throughout. Nothing crashes, nothing warns.

This is worse than the `-inf` I originally described. An `inf` loss announces itself on the first
step. A dead gradient produces a run that completes, writes a normal-looking JSON, and is simply
wrong — and the peak head's evaluation would read as "the auxiliary supervision did not help"
rather than "the auxiliary supervision was never trained".

The cliff is at logit 8 because that is where bf16's `sigmoid` reaches exactly 1.0. fp32 reaches it
only at ~40, which is why this never appeared before `--precision bf16` existed.

## The fix is unchanged and now better motivated

`binary_cross_entropy_with_logits` never forms `1 - p`, so it is exact at any precision. The same
three logits give losses 8.0003 / 12 / 40 with gradient **1.0** throughout.

## One consequence that came out of implementing it

`binary_cross_entropy_with_logits` **does not validate its target range**, where plain
`binary_cross_entropy` raises. It accepts the `MISSING = -1` sentinel and returns a finite,
plausible number. Switching to logits therefore silently deletes a guard torch had been providing
for free — the one that catches an availability-mask bug before the head trains against assays a
biosample does not have. The fix must restore that check explicitly, and the test that pinned it
must assert *our* check rather than torch's, or it passes vacuously against a function the code no
longer calls.

## Status

Implemented and verified on a MIG slice: golden still 0 ULP at the defaults, and four short training
runs (`count` fp32, `count,signal,peak` fp32 and bf16, `count,peak` bf16) all produce finite loss
curves and complete evaluation. The signal head's variance floor was fenced in the same change, for
the reason in the original report.

**Comment by mehdiforoozandeh** [comment]
Fixed in `84f0606`, on `main`.

`PeakHead` emits a logit; the loss uses `binary_cross_entropy_with_logits`, which never forms
`1 - p`. Both optional heads moved inside the fp32 fence so the Gaussian head's `+ 1e-6` variance
floor is real under bf16 rather than a rounding no-op. The target-range guard that
`binary_cross_entropy` used to provide for free is now explicit in `_elem_peak_bce`, and its test
pins our check rather than torch's.

Verified on a MIG slice:

```
old path, bf16:  logit 8/12/40 -> loss 13.8023 (frozen), gradient 0
new path, bf16:  logit 8/12/40 -> loss 8.0003 / 12 / 40, gradient 1.0000
```

plus golden 0 ULP at the defaults, 319 tests passing, and four short training runs
(`count` fp32; `count,signal,peak` fp32 and bf16; `count,peak` bf16) finite through the full
pipeline including evaluation.
